debug/gdbserver.py

```diff
@@ -562,30 +562,48 @@ def setup(self):
         self.gdb.load()
         for hart in self.target.harts:
             self.gdb.select_hart(hart)
+            self.gdb.p("$mhartid")
             self.gdb.p("$pc=_start")
 
     def test(self):
         previous_hart_count = [0 for h in self.target.harts]
         previous_interrupt_count = [0 for h in self.target.harts]
-        for _ in range(10):
-            self.gdb.c(wait=False)
-            time.sleep(2)
-            self.gdb.interrupt()
-            self.gdb.p("$mie")
-            self.gdb.p("$mip")
-            self.gdb.p("$mstatus")
-            self.gdb.p("$priv")
-            self.gdb.p("buf", fmt="")
-            hart_count = self.gdb.p("hart_count")
-            interrupt_count = self.gdb.p("interrupt_count")
-            for i, h in enumerate(self.target.harts):
-                assertGreater(hart_count[i], previous_hart_count[i])
-                assertGreater(interrupt_count[i], previous_interrupt_count[i])
-                self.gdb.select_hart(h)
-                pc = self.gdb.p("$pc")
-                self.gdb.stepi()
-                stepped_pc = self.gdb.p("$pc")
-                assertNotEqual(pc, stepped_pc)
+        # Check 10 times
+        for i in range(10):
+            # 3 attempts for each time we want the check to pass
+            for attempt in range(3):
+                self.gdb.global_command("echo round %d attempt %d\\n" % (i,
+                    attempt))
+                self.gdb.c_all(wait=False)
+                time.sleep(2)
+                self.gdb.interrupt_all()
+                hart_count = self.gdb.p("hart_count")
+                interrupt_count = self.gdb.p("interrupt_count")
+                ok = True
+                for i, h in enumerate(self.target.harts):
+                    if hart_count[i] <= previous_hart_count[i]:
+                        ok = False
+                        break
+                    if interrupt_count[i] <= previous_interrupt_count[i]:
+                        ok = False
+                        break
+                    self.gdb.p("$mie")
+                    self.gdb.p("$mip")
+                    self.gdb.p("$mstatus")
+                    self.gdb.p("$priv")
+                    self.gdb.p("buf", fmt="")
+                    self.gdb.select_hart(h)
+                    pc = self.gdb.p("$pc")
+                    self.gdb.stepi()
+                    stepped_pc = self.gdb.p("$pc")
+                    assertNotEqual(pc, stepped_pc)
+                previous_hart_count = hart_count
+                previous_interrupt_count = interrupt_count
+                if ok:
+                    break
+            else:
```

Comment:
Reading on phone so spacing may be off but what if does this else go with?

Reply:
It goes with the inner for (which does the retries if necessary).
Python allows an else clause on for/while loops, which get executed if the loop exits normally (eg. not due to break).

Reply:
TIL! Thanks.